File: JavaDatabaseCompatibility/src/com/ibm/database/compatibility/Operation.java
package com.ibm.database.compatibility;

import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.Timestamp;
import java.text.DateFormat;
import java.text.MessageFormat;
import java.text.SimpleDateFormat;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonToken;
import com.google.gson.stream.JsonWriter;

public class Operation {

	private static final Logger logger = LoggerFactory.getLogger(Operation.class);

	private String resource = null; // credential | session | statement | preparedStatement | resultSet
	private String action = null; // create | execute | fetch | close | startTransaction | commitTransaction | rollbackTransaction
	private String credentialId = null;
	private String sessionId = null;
	private String statementId = null;
	private String sql = null;
	private Integer nfetch = null;
	private Binding[] bindings = null;
	private JsonArray expectedResults = null;
	
	private Integer errorCount = null;
	private Integer line = null;
	
	private Operation() {

	}
	
	public static class Builder {
		private String resource = null;
		private String action = null;
		private String credentialId = null;
		private String sessionId = null;
		private String statementId = null;
		private String sql = null;
		private Integer nfetch = null;
		private Binding[] bindings = null;
		private JsonArray expectedResults = null;

		public Operation build() {
			Operation op = new Operation();
			op.resource = resource;
			op.action = action;
			op.credentialId = credentialId;
			op.sessionId = sessionId;
			op.statementId = statementId;
			op.sql = sql;
			op.nfetch = nfetch;
			op.bindings = bindings;
			op.expectedResults = expectedResults;
			return op;
		}

		public Builder resource(final String resource) {
			this.resource = resource;
			return this;
		}

		public Builder action(final String action) {
			this.action = action;
			return this;
		}
		
		public Builder credentialId(final String credentialId) {
			this.credentialId = credentialId;
			return this;
		}

		public Builder sessionId(final String sessionId) {
			this.sessionId = sessionId;
			return this;
		}

		public Builder statementId(final String statementId) {
			this.statementId = statementId;
			return this;
		}

		public Builder sql(final String sql) {
			this.sql = sql;
			return this;
		}
		
		public Builder nfetch(final Integer nfetch) {
			this.nfetch = nfetch;
			return this;
		}
		
		public Builder bindings(final Binding[] bindings) {
			this.bindings = new Binding[bindings.length];
			System.arraycopy(bindings, 0, this.bindings, 0, bindings.length);
			return this;
		}
		
		public Builder expectedResults(final JsonArray expectedResults) {
			this.expectedResults = expectedResults;
			return this;
		}
	}

	public void invoke(JdbcClient client) throws IOException, SQLException {
		this.errorCount = 0;
		if (resource.equalsIgnoreCase("credentials")) {
			if(action.equalsIgnoreCase("create"))	{
				/*-
				 * { "resource" : "credential" ,
				 *   "action" : "create" 
				 * }
				 * Actual server credential info should be picked up from the environment, e.g. VCAP services.
				 */
				DatabaseCredential credential = client.newCredential(credentialId);
				logger.debug("creating credential: id = {}", credential.getCredentialId());//, credential.getUrl());
			}
			if(action.equalsIgnoreCase("close")) {
				/*
				 * { "resource" : "credential" ,
				 *   "action" : "close" ,
				 *   "credentialId" : "mydb"
				 * }
				 */
				DatabaseCredential credential = client.removeCredential(credentialId);
				logger.debug("closing credential: id = {}, url = {}",credential.getCredentialId(), credential.getUrl() );
			}
		} else if (resource.equalsIgnoreCase("session")) {
			if (action.equalsIgnoreCase("create")) {
				JdbcSession session = client.newSession(sessionId);
				logger.debug("created new session id {}", session.getId());
			} else if (action.equalsIgnoreCase("close")) {
				JdbcSession session = client.removeJdbcSession(sessionId);
				logger.debug("closed session id {}", session.getId());
			} else if (action.equalsIgnoreCase("startTransaction")) {
				JdbcSession session = client.getJdbcSession(sessionId);
				session.startTransaction();
				logger.debug("starting transaction on session id {}", session.getId());
			} else if (action.equalsIgnoreCase("commitTransaction")) {
				JdbcSession session = client.getJdbcSession(sessionId);
				session.commitTransaction();
				logger.debug("committing transaction on session id {}", session.getId());
			} else if (action.equalsIgnoreCase("rollbackTransaction")) {
				JdbcSession session = client.getJdbcSession(sessionId);
				session.rollbackTransaction();
				logger.debug("rolling back transaction on session id {}", session.getId());
			} else {
				throw new RuntimeException(MessageFormat.format("Unsupported session action {0}", action));
			}
		} else if (resource.equalsIgnoreCase("statement")) {
			if (action.equalsIgnoreCase("create")) {
				/*-
				 * { "resource" : "statement" ,
				 *   "action" : "create" ,
				 *   "sessionId" : "mySession" , // optional
				 *   "statementId" : "abc" // optional
				 * } 
				 */
				JdbcSession jdbcSession = client.getJdbcSession(sessionId);
				jdbcSession.createStatement(statementId);
				logger.debug("created new statement id {}", jdbcSession.getLastStatementId());
			} else if (action.equalsIgnoreCase("execute")) {
				Statement stmt = null;
				try {
					JdbcSession jdbcSession = client.getJdbcSession(sessionId);
					stmt = jdbcSession.getStatement(statementId);
					if (stmt == null) {
						if (jdbcSession.getResultSet(statementId) != null) {
							jdbcSession.removeResultSet(statementId);
						}
						stmt = jdbcSession.createStatement(statementId);
					}
					logger.debug("executing statement (id: {} , sql: {})", jdbcSession.getLastStatementId(), sql);
					if (stmt.execute(sql)) {
						ResultSet rs = stmt.getResultSet();
						String actualResults = convertResultSetToJson(rs, nfetch);
						rs.close();
						if (expectedResults != null) {
							compareResults(expectedResults, actualResults);
						}
					}
				} finally {
					//					if (stmt != null && statementId == null) {
					//						try {
					//							stmt.close();
					//						} catch (SQLException e) {
					//							// do nothing
					//						}
					//					}
				}
			} else if (action.equalsIgnoreCase("close")) {
				JdbcSession jdbcSession = client.getJdbcSession(sessionId);
				jdbcSession.removeStatement(statementId);
				if (jdbcSession.getResultSet(statementId) != null) {
					ResultSet rs = jdbcSession.getResultSet(statementId);
					rs.close();
					jdbcSession.removeResultSet(statementId);
					logger.debug("closed result set associated with statement id {}", jdbcSession.getLastStatementId());
				}
				logger.debug("closed statement id {}", jdbcSession.getLastStatementId());
			}
		} else if (resource.equalsIgnoreCase("preparedStatement")) {
			if (action.equalsIgnoreCase("create")) {
				/*-
				 * { "resource" : "preparedStatement" ,
				 *   "action" : "create" ,
				 *   "sql" : "select tabname from systables where tabid > 99" ,
				 *   "sessionId" : "mySession" , // optional
				 *   "statementId" : "abc" // optional
				 * } 
				 */
				JdbcSession jdbcSession = client.getJdbcSession(sessionId);
				Connection c = jdbcSession.getConnection();
				PreparedStatement pstmt = c.prepareStatement(sql);
				jdbcSession.putPreparedStatement(statementId, pstmt);
				logger.debug("created prepared statement id {}", jdbcSession.getLastPreparedStatementId());
			} else if (action.equalsIgnoreCase("execute")) {
				PreparedStatement pstmt = null;
				try {
					JdbcSession jdbcSession = client.getJdbcSession(sessionId);
					pstmt = jdbcSession.getPreparedStatement(statementId);
					if (pstmt == null) {
						if (jdbcSession.getResultSet(statementId) != null) {
							jdbcSession.removeResultSet(statementId);
						}
						Connection c = jdbcSession.getConnection();
						pstmt = c.prepareStatement(sql);
						jdbcSession.putPreparedStatement(statementId, pstmt);
					}
					logger.debug("executing prepared statement (id: {})", jdbcSession.getLastPreparedStatementId());
					if (this.bindings != null) {
						Binding.bindAll(this.bindings, pstmt);
					}
					if (pstmt.execute()) {
						ResultSet rs = pstmt.getResultSet();
						String actualResults = convertResultSetToJson(rs, nfetch);
						if (nfetch == null) {
							rs.close();
						} else {
							jdbcSession.putResultSet(statementId, rs);
						}
						if (expectedResults != null) {
							compareResults(expectedResults, actualResults);
						}
					} else {
						if (expectedResults != null) {
							logError(MessageFormat.format("prepared statement (id: {0}) execution returned no results, expected {1}", statementId, expectedResults));
						}
					}
				} finally {
					//					if (pstmt != null && statementId == null) {
					//						try {
					//							pstmt.close();
					//						} catch (SQLException e) {
					//							// do nothing
					//						}
					//					}
				}
			} else if (action.equalsIgnoreCase("fetch")) {
				JdbcSession jdbcSession = client.getJdbcSession(sessionId);
				ResultSet rs = jdbcSession.getResultSet(statementId);
				if (rs == null) {
					throw new RuntimeException(MessageFormat.format("No results associated with id {0}. Cannot run fetch on an empty result", statementId));
				}
				logger.debug("executing fetch on prepared statement (id: {})", statementId);
				String actualResults = convertResultSetToJson(rs, nfetch);
				if (nfetch == null) {
					rs.close();
				}
				//System.out.println(actualResults);
				if (expectedResults != null) {
					compareResults(expectedResults, actualResults);
				}
			} else if (action.equalsIgnoreCase("close")) {
				JdbcSession jdbcSession = client.getJdbcSession(sessionId);
				jdbcSession.removeStatement(statementId);
				if (jdbcSession.getResultSet(statementId) != null) {
					ResultSet rs = jdbcSession.getResultSet(statementId);
					rs.close();
					jdbcSession.removeResultSet(statementId);
					logger.debug("closed result set associated with prepared statement id {}", jdbcSession.getLastStatementId());
				}
				logger.debug("closed prepared statement id {}", statementId);
			}
		} else {
			throw new RuntimeException(MessageFormat.format("Unsupported resource type {0}", resource));
		}
	}

	public Operation(String resource, String action, String sql) {
		this.resource = resource;
		this.action = action;
		this.sql = sql;
	}

	String convertResultSetToJson(ResultSet rs, Integer nfetch) throws IOException, SQLException {
		ResultSetMetaData rsmd = rs.getMetaData();
		StringWriter sw = new StringWriter();
		JsonWriter jw = new JsonWriter(sw);
		jw.setHtmlSafe(false);
		jw.beginArray();
		int fetched = 0;
		while ((nfetch == null || fetched < nfetch) && rs.next()) {
			jw.beginObject();
			for (int i=1; i <= rsmd.getColumnCount(); ++i) {
				jw.name(rsmd.getColumnLabel(i));
				Object value = rs.getObject(i);
				if(value instanceof java.sql.Timestamp) {
					value = rs.getTimestamp(i);
					DateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
					Timestamp ts = (java.sql.Timestamp) value;
					String columnType = rsmd.getColumnTypeName(i);
					int fractionSize = 0;
					if (columnType.contains("fraction")) {
						fractionSize = Integer.parseInt(columnType.substring(26, 27));
					}
					if (fractionSize == 0) {
						value = df.format(ts);
					} else {
						String fractionString = Integer.toString(ts.getNanos() / ((Double) Math.pow(10, 9 - fractionSize)).intValue());
						while (fractionString.length() < fractionSize) {
							fractionString = "0" + fractionString;
						}
						value = df.format(ts) + "." + fractionString;
					}
				} else if(value instanceof java.util.Date) {
					DateFormat df = new SimpleDateFormat("yyyy-MM-dd");
					value = df.format(rs.getDate(i));
				}
				GsonUtils.newGson().toJson(value, value.getClass(), jw);
			}
			jw.endObject();
			fetched++;
		}
		jw.endArray();
		jw.flush();
		String json = sw.toString();
		return json;
	}

	private void compareResults(JsonArray expectedJson, String actualJson) throws IOException {
		JsonReader expected = null;
		JsonReader actual = null;
		try {
			expected = new JsonReader(new StringReader(expectedJson.toString()));
			actual = new JsonReader(new StringReader(actualJson));
			while (expected.hasNext()) {
				JsonToken next = expected.peek();
				switch (next) {
				case BEGIN_ARRAY:
					expected.beginArray();
					try {
						actual.beginArray();
					} catch (IllegalStateException e) {
						logError(MessageFormat.format("expected begin array at at {0}", actual.getPath()));
					}
					break;
				case BEGIN_OBJECT:
					expected.beginObject();
					try {
						actual.beginObject();
					} catch (IllegalStateException e) {
						logError(MessageFormat.format("expected begin object at at {0}", actual.getPath()));
					}
					break;
				case BOOLEAN:
					try {
						if (expected.nextBoolean() != actual.nextBoolean()) {
							logError(MessageFormat.format("boolean mismatch at {0}", expected.getPath()));
						}
					} catch (IllegalStateException e) {
						logError(MessageFormat.format("expected boolean at at {0}", actual.getPath()));
					}
					break;
				case END_ARRAY:
					expected.endArray();
					try {
						actual.endArray();
					} catch (IllegalStateException e) {
						logError(MessageFormat.format("expected end array at at {0}", actual.getPath()));
					}
					break;
				case END_DOCUMENT:
					try {
						actual.hasNext();
					} catch (IllegalStateException e) {
						logError(MessageFormat.format("expected end document at at {0}", actual.getPath()));
					}
					break;
				case END_OBJECT:
					expected.endObject();
					try {
						actual.endObject();
					} catch (IllegalStateException e) {
						logError(MessageFormat.format("expected end object at at {0}", actual.getPath()));
					}
					break;
				case NAME:
					try {
						String expectedName = expected.nextName();
						String actualName = actual.nextName();
						if (!expectedName.equals(actualName)) {
							logError(MessageFormat.format("name mismatch at {0} (expected: {1}, actual: {2})", expected.getPath(), expectedName, actualName));
						}
					} catch (IllegalStateException e) {
						logError(MessageFormat.format("expected name at at {0}", actual.getPath()));
					}
					break;
				case NULL:
					expected.nextNull();
					try {
						actual.nextNull();
					} catch (IllegalStateException e) {
						logError(MessageFormat.format("expected null at at {0}", actual.getPath()));
					}
					break;
				case NUMBER:
					try {
						double expectedDouble = expected.nextDouble();
						double actualDouble = actual.nextDouble();
						if (expectedDouble != actualDouble) {
							logError(MessageFormat.format("double mismatch at {0} (expected: {1}, actual: {2})", expected.getPath(), expectedDouble, actualDouble));
						}
					} catch (IllegalStateException e) {
						logError(MessageFormat.format("expected double at at {0}", actual.getPath()));
					}
					break;
				case STRING:
					try {
						String expectedString = expected.nextString();
						String actualString = actual.nextString();
						if (!expectedString.equals(actualString)) {
							logError(MessageFormat.format("string mismatch at {0} (expected: {1}, actual: {2})", expected.getPath(), expectedString, actualString));
						}
					} catch (IllegalStateException e) {
						logError(MessageFormat.format("expected string at at {0}", actual.getPath()));
					}
					break;
				default:
					break;
				}
			}
			if (actual.hasNext()) {
				logError("compare results failed: actual has more results than expected" 
						+ "\nExpected Results --" + expectedJson.toString()
						+ "\nActual   Results --" + actualJson);
			}
		} finally {
			try {
				expected.close();
			} catch (IOException e) {
				// do nothing
			}
			try {
				actual.close();
			} catch (IOException e) {
				// do nothing
			}
		}
	}
	
	private void logError(final String message) {
		logger.error("Message: " + message);
		logger.error(toString());
		errorCount++;
	}
	
	public int getErrorCount() {
		return errorCount;
	}

	public static Operation fromJson(String line) {
		final Gson gson = GsonUtils.newGson();
		Operation operation = gson.fromJson(line, Operation.class);
		return operation;
	}

	@Override
	public String toString() {
		return GsonUtils.newGson().toJson(this);
	}

	public void setLine(int line) {
		this.line = line;
	}
	
	public int getLine() {
		return line;
	}
}

File: PHPDatabaseCompatibility/src/Operation.php
<?php
include "Binding.php";

class Operation {
	
	private $operation_json = null;
	private $resource = null; // session | statement | preparedStatement | resultSet
	private $action = null; // create | execute | fetch | close | startTransaction | commitTransaction | rollbackTransaction
	private $credentialId = null;
	private $sessionId = null;
	private $statementId = null;
	private $sql = null;
	private $nfetch = null;
	private $bindings = null;
	private $expectedResults = null;
	private $errors = array();
	private $logFile = null;
	
	function __construct($json, $logFile)  {
		$this->operation_json = $json;
		$this->resource  = (isset($json['resource'])) ? $json['resource'] : null;
		$this->action  = (isset($json['action'])) ? $json['action'] : null;
		$this->credentialId  = (isset($json['credentialId'])) ? $json['credentialId'] : null;
		$this->sessionId  = (isset($json['sessionId'])) ? $json['sessionId'] : null;
		$this->statementId  = (isset($json['statementId'])) ? $json['statementId'] : null;
		$this->className  = (isset($json['className'])) ? $json['className'] : null;
		$this->sql  = (isset($json['sql'])) ? $json['sql'] : null;
		$this->nfetch  = (isset($json['nfetch'])) ? $json['nfetch'] : null;
		if (isset($json['bindings'])) {
			$bindings_input = $json['bindings'];
			$this->bindings = array();
			$i = 0;
			foreach ($bindings_input as $b) {
				$i++;
				if (is_array($b)) {
					$this->bindings[] = new Binding($b['index'], $b['value'], $b['type']);
				} else {
					$this->bindings[] = new Binding($i, $b, null);
				}
			}
		}
		$expectedResults  = (isset($json['expectedResults'])) ? $json['expectedResults'] : null;
		if ($expectedResults != null) {
			$this->expectedResults = array();
			// Convert expected to have all caps
			foreach ($expectedResults as $row) {
				$newRow = array();
				foreach ($row as $k => $v) {
					$newRow[strtoupper($k)] = $v;
				}
				$this->expectedResults[] = $newRow;
			}	
		}
		$this->logFile = $logFile;
	}
	
	public function invoke($client) {
		if ($this->resource == "credentials") {
			if($this->action == "create") {
				$credential = $client->newCredential($this->credentialId);
				$this->logMessage("creating credential: id = {$credential->getCredentialId()}, url = {$credential->getUrl()}");
			}
			if($this->action == "close") {
				/*
				 * { "resource" : "credential" ,
				* "action" : "close" ,
				* "credentialId" : "mydb"
				* }
				*/
				$credential = $client->removeCredential($this->credentialId);
				$this->logMessage("closing credential: id =  {$credential->getCredentialId()}, url =  {$credential->getUrl()}");
			}
		} else if ($this->resource == "session") {
			if ($this->action == "create") {
				$session = $client->newSession($this->sessionId);
				$this->logMessage("created new session id " . $session->getId());
			} else if ($this->action == "close") {
				$session = $client->removePHPSession($this->sessionId);
				$this->logMessage("closed session id " .  $session->getId());
			} else if ($this->action == "startTransaction") {
				$session = $client->getPHPSession($this->sessionId);
				$conn = $session->getConnection();
				$this->logMessage("starting transaction on session id " . $session->getId());
				$ret = $conn->beginTransaction();
				if (!$ret) {
					$this->logError("Start transaction failed: " . var_export($conn->errorInfo(), true));
					throw new Exception("Could not start transaction");
				}
			} else if ($this->action == "commitTransaction") {
				$session = $client->getPHPSession($this->sessionId);
				$conn = $session->getConnection();
				$this->logMessage("committing transaction on session id " . $session->getId());
				$ret = $conn->commit();
				if (!$ret) {
					$this->logError("Commit transaction failed: " . var_export($conn->errorInfo(), true));
					throw new Exception("Could not commit transaction");
				}
			} else if ($this->action == "rollbackTransaction") {
				$session = $client->getPHPSession($this->sessionId);
				$conn = $session->getConnection();
				$this->logMessage("rolling back transaction on session id " . $session->getId());
				$ret = $conn->rollback();
				if (!$ret) {
					$this->logError("Rollback transaction failed: " . var_export($conn->errorInfo(), true));
					throw new Exception("Could not rollback transaction");
				}
			} else {
				throw new Exception("Unsupported session action " . $this->action);
			}
		} else if ($this->resource == "statement") {
			if ($this->action == "execute") {
				$session = $client->getPHPSession($this->sessionId);
				$this->logMessage("executing statement ( sql: {$this->sql})");
				if ($this->expectedResults == null) {
					$session->executeStatement($this->sql);
				} else {
					$actualResults = $session->executeQuery($this->sql, $this->nfetch);
					$this->compareResults($this->expectedResults, $actualResults);
				}
			} else {
				$this->logMessage("ignoring action {$this->action} on statement. php only supports executing a statement");
			}
		} else if ($this->resource == "preparedStatement") {
			if ($this->action == "create") {
				/*-
				 * { "resource" : "preparedStatement" ,
				* "action" : "create" ,
				* "sql" : "select tabname from systables where tabid > 99" ,
				* "sessionId" : "mySession" , // optional
				* "statementId" : "abc" // optional
				* }
				*/
				$session = $client->getPHPSession($this->sessionId);
				$c = $session->getConnection();
				$pstmt = $c->prepare($this->sql);
				$session->putPreparedStatement($this->statementId, $pstmt);
				$this->logMessage("created prepared statement id " . $session->getLastPreparedStatementId());
			} else if ($this->action == "execute") {
				$session = $client->getPHPSession($this->sessionId);
				$pstmt = $session->getPreparedStatement($this->statementId);
				if ($pstmt == null) {
					$c = $session->getConnection();
					$pstmt = $c->prepare($this->sql);
					$session->putPreparedStatement($this->statementId, $pstmt);
				}
				$this->logMessage("executing prepared statement (id: {$session->getLastPreparedStatementId()})");
				if ($this->bindings != null) {
					foreach ($this->bindings as $binding) {
						$binding->bind($pstmt);
					}
				}
				if ($pstmt->execute()) {
					$actualResults = $this->fetchRows($pstmt, $this->nfetch);
					if ($this->expectedResults != null) {
						$this->compareResults($this->expectedResults, $actualResults);
					}
				} else {
					$this->logError("FAIL: prepared statement execution returned false. " . var_export($pstmt->errorInfo(), true));
				}
			} else if ($this->action == "fetch") {
				$session = $client->getPHPSession($this->sessionId);
				$pstmt = $session->getPreparedStatement($this->statementId);
				if ($pstmt == null) {
					throw new Exception("Cannot run fetch on an empty prepared statement: statement id " . $this->statementId);
				}
				$this->logMessage("running fetch on prepared statement (id: {$this->statementId})");
				$actualResults = $this->fetchRows($pstmt, $this->nfetch);
				if ($this->expectedResults != null) {
					$this->compareResults($this->expectedResults, $actualResults);
				}
			} else if ($this->action == "close") {
				$session = $client->getPHPSession($this->sessionId);
				$session->removePreparedStatement($this->statementId);
				$this->logMessage("closed prepared statement id " . $session->getLastPreparedStatementId());
			} else {
				throw new Exception("unsupported action {$this->action} on prepared statement");
			}
		} else {
			throw new Exception("Unsupported resource type " . $this->resource);
		}
	}
	
	function fetchRows($stmt, $nfetch = null) {
		$results = array();
		if ($nfetch == null) {
			$row = $stmt->fetch(PDO::FETCH_ASSOC);
			while ($row !== false) {
				$results[] = $row;
				$row = $stmt->fetch(PDO::FETCH_ASSOC);
			}
		} else {
			for ($i = 0; $i < $nfetch; $i++) {
				$row = $stmt->fetch(PDO::FETCH_ASSOC);
				if ($row === false) {
					break;
				}
				$results[] = $row;
			}
		}
		return $results;
	}
	
	function compareResults($expectedResult, $actualResult)  {
		$comparison = $this->doCompareResults($expectedResult, $actualResult);
		if ($comparison) {
			$this->logMessage("comparing results : success");
		} else {
			$this->logError("comparing results : FAIL. Results are not equal. Expected: " . var_export($expectedResult, true) . " Actual: " . var_export($actualResult, true));
		}
		return $comparison;
	}
	
	function doCompareResults($expectedResult, $actualResult) {
		if ($expectedResult == $actualResult) {
			// if equality comparison returns true, return true
			return true;
		}
		
		// else to a more in depth comparison
		if (sizeof($expectedResult) != sizeof($actualResult)) {
			return false;
		}
		for ($i = 0; $i < sizeof($actualResult); $i++) {
			$actualRow = $actualResult[$i];
			$expectedRow = $expectedResult[$i];
			foreach ($actualRow as $name => $value) {
				if ($value != $expectedRow[$name]) {
					// Values are not equal. But there are some special checks to try.
					// If one or both numbers are floats, compare with error tolerance
					if (is_float($value) || is_float($expectedRow[$name])) {
						$precision = ini_get('precision');
 						$errorTolerance = pow(10, floor(log10($value) - ($precision - 1)));
						$diff = abs($value - $expectedRow[$name]);
						if ($diff < $errorTolerance) {
							$this->logMessage("Float comparison: values are not strictly equal, but comparison succeeding because difference is less than error tolerance");
							$this->logMessage("Float comparison details: returned value = $value, expected value = $expectedRow[$name], precision = $precision, difference = $diff, error tolerance = $errorTolerance");
						} else {
							return false;
						}
					}  else {
						// Check if values are datetime strings
						try {
							$actualDateTime = DateTime::createFromFormat("Y-m-d H:j:s.u", $value);
							$expectedDatetime = DateTime::createFromFormat((strlen($expectedRow[$name]) == 19)? "Y-m-d H:j:s" : "Y-m-d H:j:s.u", $expectedRow[$name]);
							if ($actualDateTime && $expectedDatetime) {
								// They are valid datetimes, try to compare them again after conversion to datetime objects
								if ($actualDateTime->format("Y-m-d H:j:s.u") != $expectedDatetime->format("Y-m-d H:j:s.u")) {
									// datetime objects are not equal
									return false;
								}
							} else {
								// They are not datetimes, return false (objects are not equal)
								return false;
							}
						} catch (Exception $e) {
							// Exception trying to convert object to datetime, return false (objects are not equal)
							return false;
						}
					}
				}
			} 
		}
		return true;
	}
	
	private function logMessage($msg) {
		fwrite($this->logFile, $msg . PHP_EOL);
	}

	private function logError($errmsg) {
		$this->errors[] = $errmsg;
		$this->logMessage($errmsg);
	}
	
	function getErrors() {
		return $this->errors;
	}
	
	function getErrorCount() {
		return sizeof($this->errors);
	}
	
	function __toString() {
		return json_encode($this->operation_json);
	}
}
?>